### ISSUE TITLE
Facebook Login issue - Cannot POST /login/facebook 

### ISSUE THREAD

**Issue by ndoctor13** [Azure/azure-mobile-apps-ios-client#142]
Implemented Facebook login with Mobile apps and followed all the steps according to the guidelines mentioned in this repo but when I am trying to login with facebook provider, I get an error message "Domain=com.Microsoft.MicrosoftAzureMobile.ErrorDomain Code=-1302 "Cannot POST /login/facebook
" UserInfo={NSLocalizedDescription=Cannot POST /login/facebook
})"

Tried every troubleshooting steps including the steps mentioned in this article but no luck. https://blogs.msdn.microsoft.com/appserviceteam/2017/03/30/updating-migrated-azure-mobile-services-sites-for-facebook-auth/ 

 Can someone help me as to why I am seeing this error message and how to go around it?

**Comment by bmourat** [comment]
Hi @ndoctor13 !
How are you trying to authenticate: using safari browser or webview?
Could you please share your MSClient initialization code and code for login part?
Do i understand correctly that android client is passing facebook authentication on same azure app service backend?

**Comment by ndoctor13** [comment]
Hi, We are using the facebook app to authenticate. It opens the facebook app to validate the user.
You are correct, the android client is passing facebook authentication on the same azure app with no problem. One issue I found was that the received request on the server/app was incorrect from iOS client where as it was receiving the correct URL from Android. 
Received request on app from iOS is "Received request: POST https://xyz.azurewebsites.net/login/facebook"  

While from Android is "Received request: POST https://xyz.azurewebsites.net/.auth/login/facebook" 
 
Below is our code on iOS. 

self.client = MSClient(
			applicationURLString:"https://xyz.azurewebsites.net"
		)
                
                let delegate = UIApplication.sharedApplication().delegate as? AppDelegate
                let Client = delegate!.client!;
                
                let payload: [String: String] = ["access_token": self.fbToken]
                Client.loginWithProvider("facebook", token: payload, completion: { (user, error) in
                    if user != nil{

                        NSUserDefaults.standardUserDefaults().setObject( self.fbToken, forKey: "facebookToken")
                        
                        NSUserDefaults.standardUserDefaults().setObject( user.userId, forKey: "azureUserId")
                        NSUserDefaults.standardUserDefaults().setObject( user.mobileServiceAuthenticationToken, forKey: "azureAuthenticationToken")
                        Client.currentUser = user;
                        
					} else {
						print(error)
					}
                })
		
Error is received where I call the statement "print(error)"

**Comment by dhei** [comment]
Hey @ndoctor13, Azure Mobile Apps iOS SDK doesn't support authentication using the Facebook app. When you use `loginWithProvider:` API to login, you are using `SafariViewController`-based login flow. For detailed information of how to use `loginWithProvider:` API, you may refer to https://docs.microsoft.com/en-us/azure/app-service-mobile/app-service-mobile-ios-get-started-users.

**Comment by ndoctor13** [comment]
Hi Di,

Thanks for the response. This is a little unpleasant experience for the user. I followed the steps outline in the below link which allows Facebook app to login. If Mobile apps doesn't support it then is the instructions on this link incorrect? https://docs.microsoft.com/en-us/azure/app-service-mobile/app-service-mobile-ios-how-to-use-client-library

Do I have to follow the same process for Google login i.e. via SafariViewController rather than pulling accounts from gmail app? Same thing with the link I posted above. It doesn't mention about using SafariViewController. Can you explain in what scenario can I use the link posted above?

**Comment by dhei** [comment]
Hi @ndoctor13,

Looks like there's some confusion between login with Facebook (or any other identity provider) using server-flow vs. client-flow. Let me clarify a few things.

There are two types of authentication flow, server-flow and client-flow. Server-flow is authentication flow managed by Azure App Service through a SafariViewController-based work flow. Client-flow uses an identity provider’s SDK to integrate a more native feel to the authentication flow. Basically, Azure App Service backend manage your token in the server-flow and your client doesn't have to worry about Facebook's Sign-In SDK, that's why we called it *server-flow*. While in the client-flow, your client get the token from Facebook's Sign-In SDK and login with Azure App Service backend. This article ([link](https://adrianhall.github.io/develop-mobile-apps-with-csharp-and-azure/chapter2/authconcepts/#server-side-vs-client-side)) provides a detailed explanation of server-flow and client-flow. In practice, server-flow is easier to get started with but less flexible in the long term so we recommend client-flow if you want more control of the authentication flow and better user experience. 

To answer your question, this [link](https://docs.microsoft.com/en-us/azure/app-service-mobile/app-service-mobile-ios-how-to-use-client-library#a-namefacebook-sdkahow-to-authenticate-users-with-the-facebook-sdk-for-ios) provides code snippet for client-flow using Facebook SDK. If you wish to use server-flow, you should follow this [link](https://docs.microsoft.com/en-us/azure/app-service-mobile/app-service-mobile-ios-get-started-users#a-nameadd-authenticationaadd-authentication-to-app).

Let me know if you have more questions.

**Comment by ndoctor13** [comment]
Hi Di,

I followed the exact instructions on the client-side flow using Facebook SDK per the link you provided and that is when I was getting the error (my very first comment) I reported in this issue. I get an error message on the client side indicating "Domain=com.Microsoft.MicrosoftAzureMobile.ErrorDomain Code=-1302 "Cannot POST /login/facebook
" UserInfo={NSLocalizedDescription=Cannot POST /login/facebook
})"

I am not sure the reason for this error. I looked at the server logs and noticed the received request to the server from the client was incorrect. The request received on the server is"
Received request: POST https://xyz.azurewebsites.net/login/facebook" as opposed to the required "Received request: POST https://xyz.azurewebsites.net/.auth/login/facebook ". Android client sends the correct request but iOS does not. We have followed all the steps mentioned in the link you posted but it does not seem to solve the problem. 

Thank you

**Comment by dhei** [comment]
@ndoctor13, I am having trouble reproducing this issue on my side. it is quite strange that your POST request URL looks something like "https://xyz.azurewebsites.net/login/facebook" instead of "https://xyz.azurewebsites.net/.auth/login/facebook". Can you confirm you're using the latest version (v.3.3.0) of Azure Mobile Apps iOS SDK, not any version of Azure Mobile Services SDK? I suspect you might be using old client sdk, because Azure Mobile Services indeed used "https://xyz.azure-mobile.net/login/facebook" as login URL.

**Comment by ndoctor13** [comment]
I did install the SDK correctly, I believe. I followed the instructions in this link https://github.com/Azure/azure-mobile-apps-ios-client/blob/master/README.md#ios-client-sdk

However, we were using Mobile Services SDK before so the app might be using that package instead of the newly installed Azure Mobile Apps SDK. As far as I remember, we moved all the references to Mobile Services SDK but if you're suggesting that the POST request is similar to what is returned by Mobile Services SDK then we might have not completely removed it from all places and it could be referencing it.
